File: src/agent.ts
//
// Copyright contributors to the agentic-ai-cyberres project
//
import "dotenv/config.js";
import { BeeAgent } from "bee-agent-framework/agents/bee/agent";
import { FrameworkError } from "bee-agent-framework/errors";
import { TokenMemory } from "bee-agent-framework/memory/tokenMemory";
import { OpenMeteoTool } from "bee-agent-framework/tools/weather/openMeteo";
import { getChatLLM } from "./helpers/llm.js";
import { DuckDuckGoSearchTool } from "bee-agent-framework/tools/search/duckDuckGoSearch";
import { createConsoleReader } from "./helpers/reader.js";
import { MongoDBDataValidatorTool } from "./helpers/dataValidatorTools.ts";
import { FindWhatsRunningByPortsTool } from "./helpers/dataValidatorTools.ts";
import { FindRunningProcessesTool } from "./helpers/dataValidatorTools.ts";
import { SendEmailTool } from "./helpers/dataValidatorTools.ts";

const llm = getChatLLM();
const agent = new BeeAgent({
  llm,
  memory: new TokenMemory({ llm }),
  tools: [new OpenMeteoTool(), new DuckDuckGoSearchTool(), MongoDBDataValidatorTool, FindWhatsRunningByPortsTool, FindRunningProcessesTool, SendEmailTool ],
});

const reader = createConsoleReader({ fallback: "What are the most common enterprise applications that run on Linux in the industry today?  Do not include Linux or Linux distributions in the results.  Do not identify what's currently running." });
for await (const { prompt } of reader) {
  try {
    const response = await agent
      .run(
        { prompt },
        {
          execution: {
            maxIterations: 8,
            maxRetriesPerStep: 3,
            totalMaxRetries: 10,
          },
        },
      )
      .observe((emitter) => {
        emitter.on("update", (data) => {
          reader.write(`Agent 🤖 (${data.update.key}) :`, data.update.value);
        });
      });

    reader.write(`Agent 🤖 :`, response.result.text);
  } catch (error) {
    reader.write(`Error`, FrameworkError.ensure(error).dump());
  }
}

File: python/agent.py
import asyncio
import os
from helpers.llm import get_chat_llm
from helpers.reader import ConsoleReader
from helpers.data_validator_tools import (
    MongoDBDataValidatorTool,
    FindWhatsRunningByPortsTool,
    FindRunningProcessesTool,
    SendEmailTool,
)
from bee_agent_framework import BeeAgent, TokenMemory, OpenMeteoTool, DuckDuckGoSearchTool, FrameworkError

async def main():
    llm = get_chat_llm()
    agent = BeeAgent(
        llm=llm,
        memory=TokenMemory(llm=llm),
        tools=[
            OpenMeteoTool(),
            DuckDuckGoSearchTool(),
            MongoDBDataValidatorTool(),
            FindWhatsRunningByPortsTool(),
            FindRunningProcessesTool(),
            SendEmailTool(),
        ],
    )

    reader = ConsoleReader(
        fallback="What are the most common enterprise applications that run on Linux in the industry today?  Do not include Linux or Linux distributions in the results.  Do not identify what's currently running."
    )

    async for prompt in reader:
        try:
            response = await agent.run(
                {"prompt": prompt},
                execution={
                    "maxIterations": 8,
                    "maxRetriesPerStep": 3,
                    "totalMaxRetries": 10,
                },
            )
            # Assuming response has an observe method similar to Node.js version
            async for update in response.observe():
                reader.write(f"Agent 🤖 ({update['key']}) :", update["value"])
            reader.write("Agent 🤖 :", response.result.text)
        except FrameworkError as error:
            reader.write("Error", str(error))

if __name__ == "__main__":
    import asyncio
    asyncio.run(main())
